datafusion/datasource/src/file.rs

```diff
@@ -126,19 +126,26 @@ pub trait FileSource: Send + Sync {
     /// Set optional schema adapter factory.
     ///
     /// [`SchemaAdapterFactory`] allows user to specify how fields from the
-    /// file get mapped to that of the table schema. The default implementation
-    /// returns the original source.
+    /// file get mapped to that of the table schema.  If you implement this
+    /// method, you should also implement [`schema_adapter_factory`].
     ///
-    /// Note: You can implement this method and `schema_adapter_factory`
-    /// automatically using the [`crate::impl_schema_adapter_methods`] macro.
+    /// The default implementation returns a not implemented error.
+    ///
+    /// [`schema_adapter_factory`]: Self::schema_adapter_factory
```

Comment:
this signature is changed to return a Result and defaults to not implemented

As @kosiew points out this means that implementors of a `FileSource` will only get an error at runtime, not compile time, but I think that is better than forcing everyone to implement a schema adapter even if they don't need it